File: ts/tests/conformance/encrypt_group.test.ts
import { describe, it, expect } from 'vitest';
import { encryptGroupMessage } from '../../src/v2/e2ee/encrypt-group.js';
import { decryptMessage } from '../../src/v2/e2ee/decrypt.js';
import { generateP256Keypair } from '../../src/v2/crypto/ecdh.js';
import { ProtectedHeaders } from '../../src/protected-headers.js';

const SDK_VERSION = '0.3.4';

/**
 * Group 加密自加密自解密回环：
 *   - 多成员（3DH + 1DH 混合）
 *   - state_commitment 占位 vs 显式
 */

describe('encryptGroupMessage roundtrip', () => {
  it('mixed 3DH + 1DH members', () => {
    const [aliceIkPriv, aliceIkPubDer] = generateP256Keypair();
    const [bobIkPriv, bobIkPubDer] = generateP256Keypair();
    const [bobSpkPriv, bobSpkPubDer] = generateP256Keypair();
    const [carolIkPriv, carolIkPubDer] = generateP256Keypair();

    const sender = {
      aid: 'alice.aid.com',
      deviceId: 'dev-alice-1',
      ikPriv: aliceIkPriv,
      ikPubDer: aliceIkPubDer,
    };
    const targets = [
      {
        aid: 'bob.aid.com',
        deviceId: 'dev-bob-1',
        role: 'member',
        keySource: 'group_device_prekey',
        ikPkDer: bobIkPubDer,
        spkPkDer: bobSpkPubDer,
        spkId: 'sha256:bob_spk_1',
      },
      {
        aid: 'carol.aid.com',
        deviceId: 'dev-carol-1',
        role: 'member',
        keySource: 'aid_master',
        ikPkDer: carolIkPubDer,
      },
    ];
    const payload = { text: 'group hi' };

    const envelope = encryptGroupMessage(
      sender,
      'g-test.aid.com',
      5,
      targets,
      payload,
      {},
      { state_version: 1, state_hash: 'abc', state_chain: 'chain-1' },
    );

    expect(envelope.type).toBe('e2ee.group_encrypted');
    expect(envelope.group_id).toBe('g-test.aid.com');
    expect(envelope.epoch).toBe(5);
    const aad = envelope.aad as Record<string, unknown>;
    expect(aad.wrap_protocol).toBe('1DH+3DH');
    expect(aad.state_commitment).toEqual({
      state_version: 1,
      state_hash: 'abc',
      state_chain: 'chain-1',
    });
    expect(envelope.protected_headers).toMatchObject({
      sdk_lang: 'typescript',
      sdk_version: SDK_VERSION,
    });
    expect((envelope.protected_headers as Record<string, unknown>).sdk_vesion).toBeUndefined();

    // Bob 3DH
    const decBob = decryptMessage(
      envelope as Record<string, unknown>,
      'bob.aid.com',
      'dev-bob-1',
      bobIkPriv,
      bobSpkPriv,
      aliceIkPubDer,
    );
    expect(decBob).toEqual(payload);

    // Carol 1DH
    const decCarol = decryptMessage(
      envelope as Record<string, unknown>,
      'carol.aid.com',
      'dev-carol-1',
      carolIkPriv,
      undefined,
      aliceIkPubDer,
    );
    expect(decCarol).toEqual(payload);
  });

  it('default state_commitment uses placeholder', () => {
    const [aliceIkPriv, aliceIkPubDer] = generateP256Keypair();
    const [bobIkPriv, bobIkPubDer] = generateP256Keypair();
    const sender = {
      aid: 'alice.aid.com',
      deviceId: 'dev-alice-1',
      ikPriv: aliceIkPriv,
      ikPubDer: aliceIkPubDer,
    };
    const target = {
      aid: 'bob.aid.com',
      deviceId: 'dev-bob-1',
      role: 'member',
      keySource: 'aid_master',
      ikPkDer: bobIkPubDer,
    };

    const envelope = encryptGroupMessage(sender, 'g-x', 0, [target], { hi: 1 });
    const aad = envelope.aad as Record<string, unknown>;
    expect(aad.state_commitment).toEqual({
      state_version: 0,
      state_hash: '',
      state_chain: '',
    });
    const dec = decryptMessage(
      envelope as Record<string, unknown>,
      'bob.aid.com',
      'dev-bob-1',
      bobIkPriv,
      undefined,
      aliceIkPubDer,
    );
    expect(dec).toEqual({ hi: 1 });
  });

  it('spkPkDer 存在但 spkId 为空时按 1DH 写群信封', () => {
    const [aliceIkPriv, aliceIkPubDer] = generateP256Keypair();
    const [bobIkPriv, bobIkPubDer] = generateP256Keypair();
    const [, bobSpkPubDer] = generateP256Keypair();
    const sender = {
      aid: 'alice.aid.com',
      deviceId: 'dev-alice-1',
      ikPriv: aliceIkPriv,
      ikPubDer: aliceIkPubDer,
    };
    const target = {
      aid: 'bob.aid.com',
      deviceId: 'dev-bob-1',
      role: 'member',
      keySource: 'group_device_prekey',
      ikPkDer: bobIkPubDer,
      spkPkDer: bobSpkPubDer,
      spkId: '',
    };
    const payload = { text: 'group SPK pub without ID uses 1DH' };

    const envelope = encryptGroupMessage(sender, 'g-x', 0, [target], payload);
    expect((envelope.aad as Record<string, unknown>).wrap_protocol).toBe('1DH');
    const row = (envelope.recipients as string[][])[0]!;
    expect(row[3]).toBe('aid_master');
    expect(row[5]).toBe('');
    expect(decryptMessage(
      envelope as Record<string, unknown>,
      'bob.aid.com',
      'dev-bob-1',
      bobIkPriv,
      undefined,
      aliceIkPubDer,
    )).toEqual(payload);
  });

  it('protected_headers 支持 ProtectedHeaders 实例并绑定 group V2 envelope', () => {
    const [aliceIkPriv, aliceIkPubDer] = generateP256Keypair();
    const [bobIkPriv, bobIkPubDer] = generateP256Keypair();
    const sender = {
      aid: 'alice.aid.com',
      deviceId: 'dev-alice-1',
      ikPriv: aliceIkPriv,
      ikPubDer: aliceIkPubDer,
    };
    const target = {
      aid: 'bob.aid.com',
      deviceId: 'dev-bob-1',
      role: 'member',
      keySource: 'aid_master',
      ikPkDer: bobIkPubDer,
    };

    const envelope = encryptGroupMessage(
      sender,
      'g-test.aid.com',
      2,
      [target],
      { type: 'group-text', text: 'headers' },
      {
        protectedHeaders: new ProtectedHeaders({ Device_ID: 'dev-bob-1', slot_id: 'slot-a' }),
        context: { type: 'run', id: 'run-g' },
      },
      { state_version: 7, state_hash: 'state-hash', state_chain: 'state-chain' },
    );

    expect(envelope.protected_headers).toMatchObject({
      device_id: 'dev-bob-1',
      slot_id: 'slot-a',
      payload_type: 'group-text',
    });
    expect((envelope.protected_headers as Record<string, unknown>)._auth).toBeDefined();
    expect(envelope.context).toMatchObject({ type: 'run', id: 'run-g' });
    expect((envelope.aad as Record<string, unknown>).state_commitment).toEqual({
      state_version: 7,
      state_hash: 'state-hash',
      state_chain: 'state-chain',
    });
    expect(decryptMessage(
      envelope as Record<string, unknown>,
      'bob.aid.com',
      'dev-bob-1',
      bobIkPriv,
      undefined,
      aliceIkPubDer,
    )).toEqual({ type: 'group-text', text: 'headers' });
  });

  it('未显式传 protected_headers 时也应自动注入 payload_type', () => {
    const [aliceIkPriv, aliceIkPubDer] = generateP256Keypair();
    const [bobIkPriv, bobIkPubDer] = generateP256Keypair();
    const sender = {
      aid: 'alice.aid.com',
      deviceId: 'dev-alice-1',
      ikPriv: aliceIkPriv,
      ikPubDer: aliceIkPubDer,
    };
    const target = {
      aid: 'bob.aid.com',
      deviceId: 'dev-bob-1',
      role: 'member',
      keySource: 'aid_master',
      ikPkDer: bobIkPubDer,
    };

    const envelope = encryptGroupMessage(
      sender,
      'g-test.aid.com',
      2,
      [target],
      { type: 'group-text', text: 'default headers' },
      {},
      { state_version: 7, state_hash: 'state-hash', state_chain: 'state-chain' },
    );

    expect(envelope.protected_headers).toMatchObject({
      payload_type: 'group-text',
    });
    expect((envelope.protected_headers as Record<string, unknown>)._auth).toBeDefined();
    expect(decryptMessage(
      envelope as Record<string, unknown>,
      'bob.aid.com',
      'dev-bob-1',
      bobIkPriv,
      undefined,
      aliceIkPubDer,
    )).toEqual({ type: 'group-text', text: 'default headers' });
  });

  it('group payload.type 应复制到信封顶层 payload_type', () => {
    const [aliceIkPriv, aliceIkPubDer] = generateP256Keypair();
    const [bobIkPriv, bobIkPubDer] = generateP256Keypair();
    const sender = {
      aid: 'alice.aid.com',
      deviceId: 'dev-alice-1',
      ikPriv: aliceIkPriv,
      ikPubDer: aliceIkPubDer,
    };
    const target = {
      aid: 'bob.aid.com',
      deviceId: 'dev-bob-1',
      role: 'member',
      keySource: 'aid_master',
      ikPkDer: bobIkPubDer,
    };

    const envelope = encryptGroupMessage(
      sender,
      'g-test.aid.com',
      2,
      [target],
      { type: 'group-text', text: 'visible group type' },
    );

    expect(envelope.payload_type).toBe('group-text');
    expect((envelope.protected_headers as Record<string, unknown>).payload_type).toBe('group-text');
    expect((envelope.protected_headers as Record<string, unknown>).sdk_lang).toBe('typescript');
    expect((envelope.protected_headers as Record<string, unknown>).sdk_version).toBe(SDK_VERSION);
    expect((envelope.protected_headers as Record<string, unknown>).sdk_vesion).toBeUndefined();
    expect(decryptMessage(
      envelope as Record<string, unknown>,
      'bob.aid.com',
      'dev-bob-1',
      bobIkPriv,
      undefined,
      aliceIkPubDer,
    )).toEqual({ type: 'group-text', text: 'visible group type' });
  });
});

File: go/v2/e2ee/encrypt_group_test.go
package e2ee

import (
	"reflect"
	"testing"

	"github.com/modelunion/aun-sdk-core/go/v2/crypto"
)

// TestEncryptGroupRoundtrip3DH 单成员 3DH 群消息 round-trip。
func TestEncryptGroupRoundtrip3DH(t *testing.T) {
	sender := makeTestSender(t)
	target, ikPriv, spkPriv := makeTestRecipient(t, "member", "group_device_prekey", true)

	payload := map[string]any{
		"text": "Group hello",
		"type": "group_text",
	}

	envelope, err := EncryptGroupMessage(sender, "g-test.aid.com", 5,
		[]Target{target}, payload, EncryptOptions{}, &StateCommitmentAAD{
			StateVersion: 1,
			StateHash:    "abc123",
			StateChain:   "chain-link-1",
		})
	if err != nil {
		t.Fatalf("加密失败: %v", err)
	}

	if envelope["type"] != "e2ee.group_encrypted" {
		t.Fatalf("type 字段错误: %v", envelope["type"])
	}
	if envelope["group_id"] != "g-test.aid.com" {
		t.Fatalf("group_id 错误: %v", envelope["group_id"])
	}
	if envelope["epoch"].(int) != 5 {
		t.Fatalf("epoch 错误: %v", envelope["epoch"])
	}

	aad := envelope["aad"].(map[string]any)
	if aad["wrap_protocol"] != "3DH" {
		t.Fatalf("wrap_protocol 错误: %v", aad["wrap_protocol"])
	}
	sc := aad["state_commitment"].(map[string]any)
	if sc["state_version"] != 1 {
		t.Fatalf("state_version 错误: %v", sc["state_version"])
	}

	decrypted, err := DecryptMessage(envelope, target.AID, target.DeviceID, ikPriv, spkPriv, sender.IKPubDER)
	if err != nil {
		t.Fatalf("解密失败: %v", err)
	}
	if !reflect.DeepEqual(decrypted, payload) {
		t.Fatalf("payload 不一致\n期望: %v\n实际: %v", payload, decrypted)
	}
}

func TestEncryptGroupSPKPublicKeyWithoutIDUses1DH(t *testing.T) {
	sender := makeTestSender(t)
	target, ikPriv, _ := makeTestRecipient(t, "member", "group_device_prekey", true)
	target.SPKID = ""
	payload := map[string]any{"text": "group SPK pub without ID"}

	envelope, err := EncryptGroupMessage(sender, "g-test.aid.com", 5,
		[]Target{target}, payload, EncryptOptions{}, nil)
	if err != nil {
		t.Fatalf("加密失败: %v", err)
	}
	if got := envelope["aad"].(map[string]any)["wrap_protocol"]; got != "1DH" {
		t.Fatalf("wrap_protocol 应为 1DH，实际: %v", got)
	}
	rows, err := convertRecipientsToStringMatrix(envelope["recipients"])
	if err != nil {
		t.Fatal(err)
	}
	if rows[0][3] != "aid_master" || rows[0][5] != "" {
		t.Fatalf("1DH row 应写 aid_master/空 spk_id，实际: %v", rows[0])
	}
	decrypted, err := DecryptMessage(envelope, target.AID, target.DeviceID, ikPriv, nil, sender.IKPubDER)
	if err != nil {
		t.Fatalf("1DH 解密失败: %v", err)
	}
	if !reflect.DeepEqual(decrypted, payload) {
		t.Fatalf("payload 不一致\n期望: %v\n实际: %v", payload, decrypted)
	}
}

func TestEncryptGroupPayloadTypeTopLevel(t *testing.T) {
	sender := makeTestSender(t)
	target, ikPriv, _ := makeTestRecipient(t, "member", "aid_master", false)
	payload := map[string]any{"type": "group-text", "text": "visible group type"}

	envelope, err := EncryptGroupMessage(sender, "g-test.aid.com", 5, []Target{target}, payload, EncryptOptions{}, nil)
	if err != nil {
		t.Fatalf("加密失败: %v", err)
	}
	if got := envelope["payload_type"]; got != "group-text" {
		t.Fatalf("payload_type 应在群信封顶层透传，实际: %#v", got)
	}
	headers, ok := envelope["protected_headers"].(map[string]any)
	if !ok {
		t.Fatalf("protected_headers 应存在，实际: %#v", envelope["protected_headers"])
	}
	if got := headers["payload_type"]; got != "group-text" {
		t.Fatalf("protected_headers.payload_type 应保留兼容校验，实际: %#v", got)
	}
	if headers["sdk_lang"] != e2eeSDKLang || headers["sdk_version"] != e2eeSDKVersion {
		t.Fatalf("protected_headers SDK 元信息错误: %#v", headers)
	}
	if _, ok := headers["sdk_vesion"]; ok {
		t.Fatalf("protected_headers 不应包含历史拼写 sdk_vesion: %#v", headers)
	}
	decrypted, err := DecryptMessage(envelope, target.AID, target.DeviceID, ikPriv, nil, sender.IKPubDER)
	if err != nil {
		t.Fatalf("解密失败: %v", err)
	}
	if !reflect.DeepEqual(decrypted, payload) {
		t.Fatalf("payload 不一致\n期望: %v\n实际: %v", payload, decrypted)
	}
}

// TestEncryptGroupMixedProtocol 群多成员（3DH + 1DH）。
func TestEncryptGroupMixedProtocol(t *testing.T) {
	sender := makeTestSender(t)

	bob, bobIK, bobSPK := makeTestRecipient(t, "member", "group_device_prekey", true)

	carolIK, carolPub, err := crypto.GenerateP256Keypair()
	if err != nil {
		t.Fatalf("生成 carol 失败: %v", err)
	}
	carol := Target{
		AID:       "carol.aid.com",
		DeviceID:  "dev-carol-1",
		Role:      "member",
		KeySource: "aid_master",
		IKPkDER:   carolPub,
	}

	payload := map[string]any{
		"text": "Group message",
		"type": "group_text",
	}

	envelope, err := EncryptGroupMessage(sender, "g-mixed.aid.com", 7,
		[]Target{bob, carol}, payload, EncryptOptions{}, nil)
	if err != nil {
		t.Fatalf("加密失败: %v", err)
	}

	if got := envelope["aad"].(map[string]any)["wrap_protocol"]; got != "1DH+3DH" {
		t.Fatalf("wrap_protocol 错误: %v", got)
	}

	// 默认占位 state_commitment
	sc := envelope["aad"].(map[string]any)["state_commitment"].(map[string]any)
	if sc["state_version"] != 0 || sc["state_hash"] != "" || sc["state_chain"] != "" {
		t.Fatalf("默认 state_commitment 应全 0/空: %v", sc)
	}

	// Bob 解密
	gotBob, err := DecryptMessage(envelope, bob.AID, bob.DeviceID, bobIK, bobSPK, sender.IKPubDER)
	if err != nil {
		t.Fatalf("Bob 解密失败: %v", err)
	}
	if !reflect.DeepEqual(gotBob, payload) {
		t.Fatalf("Bob payload 不一致")
	}

	// Carol 解密（1DH，无 SPK）
	gotCarol, err := DecryptMessage(envelope, carol.AID, carol.DeviceID, carolIK, nil, sender.IKPubDER)
	if err != nil {
		t.Fatalf("Carol 解密失败: %v", err)
	}
	if !reflect.DeepEqual(gotCarol, payload) {
		t.Fatalf("Carol payload 不一致")
	}
}

// TestEncryptGroupSignatureVerifies 群消息签名验证。
func TestEncryptGroupSignatureVerifies(t *testing.T) {
	sender := makeTestSender(t)
	target, _, _ := makeTestRecipient(t, "member", "aid_master", false)

	envelope, err := EncryptGroupMessage(sender, "g.aid.com", 1,
		[]Target{target}, map[string]any{"x": "y"}, EncryptOptions{}, nil)
	if err != nil {
		t.Fatalf("加密失败: %v", err)
	}
	if err := verifySenderSignature(envelope, sender.IKPubDER); err != nil {
		t.Fatalf("签名验证失败: %v", err)
	}
}
